Upload JSON Files/No selection error message

AbnerSD opened the pull request:
We added the JSON upload file containers to replace "all_contests_bristol" and "reported_results_bristol".
We made changes to conductor.py, index.html, and rla_ui.js.

**conductor.py**
- Added Python functions to take both .json file uploads. They check for .json format and throw an error if the file doesn't contain the right word dictionary. 

**index.html**
- Added the upload forms for both of the new upload .json containers

**rla_ui.js**
- Added javascript functions to display the upload container from index.html. 
- Changed the order to go from auditType to uploading both .json files first. 

**Files for testing .json upload below**
- First .json - https://www.mediafire.com/file/s7eepydx8f5o4rl/firstUpload.json/file
- Second .json - https://www.mediafire.com/file/d6t1f2d99vtmi7k/secondUpload.json/file

AbnerSD committed:
Update conductor.py

AbnerSD committed:
Update conductor.py

AbnerSD committed:
Update index.html

AbnerSD committed:
Added JSON upload container

AbnerSD committed:
Update conductor.py

AbnerSD committed:
Update rla_ui.js

AbnerSD committed:
Update index.html

AbnerSD committed:
Update rla_ui.js

AbnerSD committed:
Update conductor.py

gam22 committed:
Update conductor.py

AbnerSD committed:
Merge pull request #1 from gam22/master

Updated with .json upload python functions

AbnerSD committed:
Update conductor.py

AbnerSD committed:
Update conductor.py

tom-votes commented:
Thank you! This looks quite good. Adding a few inline comments:

tom-votes commented on conductor.py:
```diff
@@ -120,311 +122,13 @@ def add_non_candidate_choices(l):
-       'candidates': ['Approve', 'Reject'],
-      },
-      ]
+all_contests_bristol = []
```
We don't need `_bristol` in the names now

tom-votes commented on conductor.py:
```diff
@@ -985,6 +689,99 @@ def log_time():
 
+# JSON file upload function to allow first .json upload and check text format 
+# It checks "id:" "title:" and "candidates:"
+@app.route('/upload-ballot-manifest-json', methods=['POST'])
```
This isn't a ballot manifest, so a better name might be something like `'/upload-all-contests-data'`

tom-votes commented:
Also as a side note, can you make the commit messages be a bit more descriptive? E.g. your commit message "Added JSON upload container" is more helpful when going back through the git history than "Update conductor.py". Thanks!

tom-votes commented on conductor.py:
```diff
@@ -985,6 +689,99 @@ def log_time():
+
+            global default_audit_state
+            # updating default_audit_state key that depend on all_contests_bristol
+            default_audit_state['all_contests']['ballot_comparison'] = all_contests_bristol
```
This `'ballot_comparison'` may not be accurate -- we want to be able to upload this file for any contest type. Can we instead send the audit type (which will normally be either `'ballot_polling'`, `'ballot_comparison`, or `'ri_pilot'` with the form upload request?

tom-votes replied:
Let me know if I need to provide more detail about this one!

tom-votes commented on conductor.py:
```diff
@@ -985,6 +689,99 @@ def log_time():
+
+            global audit_types
+            # updating audit_types key that depend on all_contests_bristol
+            audit_types['ballot_comparison']['all_contests'] = all_contests_bristol
```
Same note about `'ballot_comparison'`

tom-votes commented on conductor.py:
```diff
@@ -985,6 +689,99 @@ def log_time():
+            audit_types['ballot_comparison']['all_contests'] = all_contests_bristol
+
+            global audit_state
+            audit_state = copy.deepcopy(default_audit_state)
```
Why are we resetting to `default_audit_state` here? That'll reset all of our state; do we want that?

tom-votes commented on conductor.py:
```diff
@@ -985,6 +689,99 @@ def log_time():
+# JSON function end
+
+# Second Upload .json File Function
+@app.route('/upload-ballot-manifest-json2', methods=['POST'])
```
Same note about naming; I'd recommend something like `/upload-reported-results`

tom-votes commented on ui/index.html:
```diff
@@ -7,7 +7,15 @@
    <body>
-      <div id="auditTypeContainer" class="container">
+     <!-- JSON file upload --> 
+	  <div id="jsonFileUploadContainer" class="container" style="display:block">
```
`"allContestDataUploadContainer"`, etc?

tom-votes commented on ui/rla_ui.js:
```diff
@@ -1014,4 +1025,253 @@ function restrictInput(event) {
+function maybeGetJSON2File(ballotType) {
+    return function() {
+        document.body.appendChild(json2FileUploadContainer);
+        if (conductorState['json_hash'][ballotType] === undefined) {
```
We don't set this anywhere yet, right? So this is for the future?

tom-votes commented on conductor.py:
```diff
@@ -985,6 +689,99 @@ def log_time():
+        try:
+            global reported_results_bristol
+            # reading it as json
+            uploaded_json = json.loads(json.dumps(ast.literal_eval(uploaded_content)))
```
How does this differ from `json.loads`? If we're uploading JSON do we need `literal_eval`?

tom-votes commented on ui/rla_ui.js:
```diff
@@ -165,9 +169,12 @@ function maybeGetAuditType(andThen) {
+       
       } else {
-         displayAuditType();
+         displayJSONFile();
```
I think these each need their own `if ( ! uiState['got_audit_type']) {`-style if-statement. The idea is that if we reload the page it'll bring us to the right spot (and we won't need to restart our work).

tom-votes commented on ui/rla_ui.js:
```diff
@@ -1014,4 +1025,253 @@ function restrictInput(event) {
+  makeBallotManifestUploadContainerJson(ballotType);
+};
+ // Creates front end form Container for JSON uploading
+function makeBallotManifestUploadContainerJson(ballotType) {
```
This is pretty similar to `makeBallotManifestUploadContainer` - can we make a general function that works for all three uploads, and takes as arguments any differences?

tom-votes commented:
@AbnerSD @gam22 that's my comments for now. Thanks, this is no small feature and it's looking pretty good!

gam22 committed:
Updated CSS for no choice selection message 

Styling for no choice selection text.

gam22 committed:
Added JavaScript to display "hidden message" 

We loop trough the array of questions generated to display an error message when not all of the choice questions are checked!

AbnerSD committed:
Merge pull request #2 from gam22/master

Error Message when not all choices are made

AbnerSD committed:
Update rla_ui.js
